Support range definition for response codes.

k1LoW opened the pull request:
Hi all.

This fix supports the range definition for response codes.

ref: https://swagger.io/docs/specification/describing-responses/

k1LoW commented:
Hi @daveshanley 
`codecov/project` is FAIL, is there anything I need to do additionally to address this?

daveshanley commented:
Hi @k1LoW yeah, there are some indirect changes that prevent some code from executing

https://app.codecov.io/gh/pb33f/libopenapi-validator/pull/65/blob/responses/validate_response.go

As you can see there are two branches that now no-longer run, would you be able to address this?

Thanks

k1LoW commented:
@daveshanley 

> Report is 9 commits behind head on main. ( ref: https://github.com/pb33f/libopenapi-validator/pull/65#issuecomment-2058200104 )

The current main branch is [b3f6e33](https://github.com/pb33f/libopenapi-validator/commit/b3f6e333ad8dce941d61602161c400991721eae4).
Comparing coverage with it, the coverage range seems to be the same.

Are you saying that there is a problem with the commitments already taken in?

daveshanley commented:
You can see in this PR for example: https://github.com/pb33f/libopenapi/pull/274 I am battling the code coverage over time after fixing bugs and regressions, sometimes the indirect coverage drop is a head scratcher - but it means that code that was running before, is no-longer covered by a test.

daveshanley commented:
> Hi @k1LoW yeah, there are some indirect changes that prevent some code from executing
> 
> https://app.codecov.io/gh/pb33f/libopenapi-validator/pull/65/blob/responses/validate_response.go
> 
> As you can see there are two branches that now no-longer run, would you be able to address this?
> 
> Thanks

What I mean by branches is 'code paths' vs git branches, there are two blocks of code that are marked red, which means they are not covered by any testing anymore.

k1LoW commented:
It is already not covered two branches by the `main`.

ref: https://app.codecov.io/gh/pb33f/libopenapi-validator/blob/main/responses%2Fvalidate_response.go

<details>

The number in the first line is the line number and the number in the second line is the cover count.
A line without a number on the second line indicates that it is not covered by the test.

```console
# /src/github.com/pb33f/libopenapi-validator (main)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.781s  coverage: 51.9% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        0.600s  coverage: 91.0% of statements
ok      github.com/pb33f/libopenapi-validator/paths     0.897s  coverage: 36.8% of statements
ok      github.com/pb33f/libopenapi-validator/requests  0.314s  coverage: 36.0% of statements
ok      github.com/pb33f/libopenapi-validator/responses 1.072s  coverage: 40.7% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 1.229s  coverage: 33.9% of statements
# /src/github.com/pb33f/libopenapi-validator (main)> octocov cat responses/validate_response.go
  1    // Copyright 2023 Princess B33f Heavy Industries / Dave Shanley
  2    // SPDX-License-Identifier: MIT
  3
  4    package responses
  5
  6    import (
  7     "bytes"
  8     "encoding/json"
  9     "fmt"
 10     "github.com/pb33f/libopenapi-validator/errors"
 11     "github.com/pb33f/libopenapi-validator/helpers"
 12     "github.com/pb33f/libopenapi-validator/schema_validation"
 13     "github.com/pb33f/libopenapi/datamodel/high/base"
 14     "github.com/santhosh-tekuri/jsonschema/v5"
 15     "gopkg.in/yaml.v3"
 16     "io"
 17     "net/http"
 18     "reflect"
 19     "regexp"
 20     "strconv"
 21     "strings"
 22    )
 23
 24    var instanceLocationRegex = regexp.MustCompile(`^/(\d+)`)
 25
 26    // ValidateResponseSchema will validate the response body for a http.Response pointer. The request is used to
 27    // locate the operation in the specification, the response is used to ensure the response code, media type and the
 28    // schema of the response body are valid.
 29    //
 30    // This function is used by the ValidateResponseBody function, but can be used independently.
 31    func ValidateResponseSchema(
 32     request *http.Request,
 33     response *http.Response,
 34     schema *base.Schema,
 35     renderedSchema,
 36 12  jsonSchema []byte) (bool, []*errors.ValidationError) {
 37 12
 38 12  var validationErrors []*errors.ValidationError
 39 12
 40 12  if response == nil || response.Body == nil {
 41             // cannot decode the response body, so it's not valid
 42             violation := &errors.SchemaValidationFailure{
 43                     Reason:          "response is empty",
 44                     Location:        "unavailable",
 45                     ReferenceSchema: string(renderedSchema),
 46             }
 47             validationErrors = append(validationErrors, &errors.ValidationError{
 48                     ValidationType:    "response",
 49                     ValidationSubType: "object",
 50                     Message: fmt.Sprintf("%s response object is missing for '%s'",
 51                             request.Method, request.URL.Path),
 52                     Reason:                 fmt.Sprintf("The response object is completely missing"),
 53                     SpecLine:               1,
 54                     SpecCol:                0,
 55                     SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
 56                     HowToFix:               "ensure response object has been set",
 57                     Context:                string(renderedSchema), // attach the rendered schema to the error
 58             })
 59             return false, validationErrors
 60     }
 61
 62 12  responseBody, ioErr := io.ReadAll(response.Body)
 63 12  if ioErr != nil {
 64             // cannot decode the response body, so it's not valid
 65             violation := &errors.SchemaValidationFailure{
 66                     Reason:          ioErr.Error(),
 67                     Location:        "unavailable",
 68                     ReferenceSchema: string(renderedSchema),
 69                     ReferenceObject: string(responseBody),
 70             }
 71             validationErrors = append(validationErrors, &errors.ValidationError{
 72                     ValidationType:    helpers.ResponseBodyValidation,
 73                     ValidationSubType: helpers.Schema,
 74                     Message: fmt.Sprintf("%s response body for '%s' cannot be read, it's empty or malformed",
 75                             request.Method, request.URL.Path),
 76                     Reason:                 fmt.Sprintf("The response body cannot be decoded: %s", ioErr.Error()),
 77                     SpecLine:               1,
 78                     SpecCol:                0,
 79                     SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
 80                     HowToFix:               "ensure body is not empty",
 81                     Context:                string(renderedSchema), // attach the rendered schema to the error
 82             })
 83             return false, validationErrors
 84     }
 85
 86     // close the request body, so it can be re-read later by another player in the chain
 87 12  _ = response.Body.Close()
 88 12  response.Body = io.NopCloser(bytes.NewBuffer(responseBody))
 89 12
 90 12  var decodedObj interface{}
 91 12
 92 22  if len(responseBody) > 0 {
 93 10          err := json.Unmarshal(responseBody, &decodedObj)
 94 10
 95 11          if err != nil {
 96  1                  // cannot decode the response body, so it's not valid
 97  1                  violation := &errors.SchemaValidationFailure{
 98  1                          Reason:          err.Error(),
 99  1                          Location:        "unavailable",
100  1                          ReferenceSchema: string(renderedSchema),
101  1                          ReferenceObject: string(responseBody),
102  1                  }
103  1                  validationErrors = append(validationErrors, &errors.ValidationError{
104  1                          ValidationType:    helpers.ResponseBodyValidation,
105  1                          ValidationSubType: helpers.Schema,
106  1                          Message: fmt.Sprintf("%s response body for '%s' failed to validate schema",
107  1                                  request.Method, request.URL.Path),
108  1                          Reason:                 fmt.Sprintf("The response body cannot be decoded: %s", err.Error()),
109  1                          SpecLine:               1,
110  1                          SpecCol:                0,
111  1                          SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
112  1                          HowToFix:               errors.HowToFixInvalidSchema,
113  1                          Context:                string(renderedSchema), // attach the rendered schema to the error
114  1                  })
115  1                  return false, validationErrors
116  1          }
117     }
118
119     // no response body? failed to decode anything? nothing to do here.
120 13  if responseBody == nil || decodedObj == nil {
121  2          return true, nil
122  2  }
123
124     // create a new jsonschema compiler and add in the rendered JSON schema.
125  9  compiler := jsonschema.NewCompiler()
126  9  fName := fmt.Sprintf("%s.json", helpers.ResponseBodyValidation)
127  9  _ = compiler.AddResource(fName,
128  9          strings.NewReader(string(jsonSchema)))
129  9  jsch, _ := compiler.Compile(fName)
130  9
131  9  // validate the object against the schema
132  9  scErrs := jsch.Validate(decodedObj)
133 14  if scErrs != nil {
134  5          jk := scErrs.(*jsonschema.ValidationError)
135  5
136  5          // flatten the validationErrors
137  5          schFlatErrs := jk.BasicOutput().Errors
138  5          var schemaValidationErrors []*errors.SchemaValidationFailure
139 21          for q := range schFlatErrs {
140 16                  er := schFlatErrs[q]
141 21                  if er.KeywordLocation == "" || strings.HasPrefix(er.Error, "doesn't validate with") {
142  5                          continue // ignore this error, it's useless tbh, utter noise.
143                     }
144 22                  if er.Error != "" {
145 11
146 11                          // re-encode the schema.
147 11                          var renderedNode yaml.Node
148 11                          _ = yaml.Unmarshal(renderedSchema, &renderedNode)
149 11
150 11                          // locate the violated property in the schema
151 11                          located := schema_validation.LocateSchemaPropertyNodeByJSONPath(renderedNode.Content[0], er.KeywordLocation)
152 11
153 11                          // extract the element specified by the instance
154 11                          val := instanceLocationRegex.FindStringSubmatch(er.InstanceLocation)
155 11                          var referenceObject string
156 11
157 13                          if len(val) > 0 {
158  2                                  referenceIndex, _ := strconv.Atoi(val[1])
159  4                                  if reflect.ValueOf(decodedObj).Type().Kind() == reflect.Slice {
160  2                                          found := decodedObj.([]any)[referenceIndex]
161  2                                          recoded, _ := json.MarshalIndent(found, "", "  ")
162  2                                          referenceObject = string(recoded)
163  2                                  }
164                             }
165 20                          if referenceObject == "" {
166  9                                  referenceObject = string(responseBody)
167  9                          }
168
169 11                          violation := &errors.SchemaValidationFailure{
170 11                                  Reason:          er.Error,
171 11                                  Location:        er.KeywordLocation,
172 11                                  ReferenceSchema: string(renderedSchema),
173 11                                  ReferenceObject: referenceObject,
174 11                                  OriginalError:   jk,
175 11                          }
176 11                          // if we have a location within the schema, add it to the error
177 22                          if located != nil {
178 11
179 11                                  line := located.Line
180 11                                  // if the located node is a map or an array, then the actual human interpretable
181 11                                  // line on which the violation occurred is the line of the key, not the value.
182 14                                  if located.Kind == yaml.MappingNode || located.Kind == yaml.SequenceNode {
183  6                                          if line > 0 {
184  3                                                  line--
185  3                                          }
186                                     }
187
188                                     // location of the violation within the rendered schema.
189 11                                  violation.Line = line
190 11                                  violation.Column = located.Column
191                             }
192 11                          schemaValidationErrors = append(schemaValidationErrors, violation)
193                     }
194             }
195
196  5          line := 1
197  5          col := 0
198 10          if schema.GoLow().Type.KeyNode != nil {
199  5                  line = schema.GoLow().Type.KeyNode.Line
200  5                  col = schema.GoLow().Type.KeyNode.Column
201  5          }
202
203             // add the error to the list
204  5          validationErrors = append(validationErrors, &errors.ValidationError{
205  5                  ValidationType:    helpers.ResponseBodyValidation,
206  5                  ValidationSubType: helpers.Schema,
207  5                  Message: fmt.Sprintf("%d response body for '%s' failed to validate schema",
208  5                          response.StatusCode, request.URL.Path),
209  5                  Reason: fmt.Sprintf("The response body for status code '%d' is defined as an object. "+
210  5                          "However, it does not meet the schema requirements of the specification", response.StatusCode),
211  5                  SpecLine:               line,
212  5                  SpecCol:                col,
213  5                  SchemaValidationErrors: schemaValidationErrors,
214  5                  HowToFix:               errors.HowToFixInvalidSchema,
215  5                  Context:                string(renderedSchema), // attach the rendered schema to the error
216  5          })
217     }
218 14  if len(validationErrors) > 0 {
219  5          return false, validationErrors
220  5  }
221  4  return true, nil
222    }
# /src/github.com/pb33f/libopenapi-validator (main)>
# /src/github.com/pb33f/libopenapi-validator (main)> git switch fix-validate-security
Switched to branch 'fix-validate-security'
# /src/github.com/pb33f/libopenapi-validator (fix-validate-security)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.941s  coverage: 50.0% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        1.506s  coverage: 90.7% of statements
ok      github.com/pb33f/libopenapi-validator/paths     0.404s  coverage: 35.5% of statements
ok      github.com/pb33f/libopenapi-validator/requests  0.580s  coverage: 33.7% of statements
ok      github.com/pb33f/libopenapi-validator/responses 0.994s  coverage: 39.6% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 1.239s  coverage: 34.3% of statements
# /src/github.com/pb33f/libopenapi-validator (fix-validate-security)> octocov cat responses/validate_response.go
  1    // Copyright 2023 Princess B33f Heavy Industries / Dave Shanley
  2    // SPDX-License-Identifier: MIT
  3
  4    package responses
  5
  6    import (
  7     "bytes"
  8     "encoding/json"
  9     "fmt"
 10     "github.com/pb33f/libopenapi-validator/errors"
 11     "github.com/pb33f/libopenapi-validator/helpers"
 12     "github.com/pb33f/libopenapi-validator/schema_validation"
 13     "github.com/pb33f/libopenapi/datamodel/high/base"
 14     "github.com/santhosh-tekuri/jsonschema/v5"
 15     "gopkg.in/yaml.v3"
 16     "io"
 17     "net/http"
 18     "reflect"
 19     "regexp"
 20     "strconv"
 21     "strings"
 22    )
 23
 24    var instanceLocationRegex = regexp.MustCompile(`^/(\d+)`)
 25
 26    // ValidateResponseSchema will validate the response body for a http.Response pointer. The request is used to
 27    // locate the operation in the specification, the response is used to ensure the response code, media type and the
 28    // schema of the response body are valid.
 29    //
 30    // This function is used by the ValidateResponseBody function, but can be used independently.
 31    func ValidateResponseSchema(
 32     request *http.Request,
 33     response *http.Response,
 34     schema *base.Schema,
 35     renderedSchema,
 36 12  jsonSchema []byte) (bool, []*errors.ValidationError) {
 37 12
 38 12  var validationErrors []*errors.ValidationError
 39 12
 40 12  if response == nil || response.Body == nil {
 41             // cannot decode the response body, so it's not valid
 42             violation := &errors.SchemaValidationFailure{
 43                     Reason:          "response is empty",
 44                     Location:        "unavailable",
 45                     ReferenceSchema: string(renderedSchema),
 46             }
 47             validationErrors = append(validationErrors, &errors.ValidationError{
 48                     ValidationType:    "response",
 49                     ValidationSubType: "object",
 50                     Message: fmt.Sprintf("%s response object is missing for '%s'",
 51                             request.Method, request.URL.Path),
 52                     Reason:                 fmt.Sprintf("The response object is completely missing"),
 53                     SpecLine:               1,
 54                     SpecCol:                0,
 55                     SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
 56                     HowToFix:               "ensure response object has been set",
 57                     Context:                string(renderedSchema), // attach the rendered schema to the error
 58             })
 59             return false, validationErrors
 60     }
 61
 62 12  responseBody, ioErr := io.ReadAll(response.Body)
 63 12  if ioErr != nil {
 64             // cannot decode the response body, so it's not valid
 65             violation := &errors.SchemaValidationFailure{
 66                     Reason:          ioErr.Error(),
 67                     Location:        "unavailable",
 68                     ReferenceSchema: string(renderedSchema),
 69                     ReferenceObject: string(responseBody),
 70             }
 71             validationErrors = append(validationErrors, &errors.ValidationError{
 72                     ValidationType:    helpers.ResponseBodyValidation,
 73                     ValidationSubType: helpers.Schema,
 74                     Message: fmt.Sprintf("%s response body for '%s' cannot be read, it's empty or malformed",
 75                             request.Method, request.URL.Path),
 76                     Reason:                 fmt.Sprintf("The response body cannot be decoded: %s", ioErr.Error()),
 77                     SpecLine:               1,
 78                     SpecCol:                0,
 79                     SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
 80                     HowToFix:               "ensure body is not empty",
 81                     Context:                string(renderedSchema), // attach the rendered schema to the error
 82             })
 83             return false, validationErrors
 84     }
 85
 86     // close the request body, so it can be re-read later by another player in the chain
 87 12  _ = response.Body.Close()
 88 12  response.Body = io.NopCloser(bytes.NewBuffer(responseBody))
 89 12
 90 12  var decodedObj interface{}
 91 12
 92 22  if len(responseBody) > 0 {
 93 10          err := json.Unmarshal(responseBody, &decodedObj)
 94 10
 95 11          if err != nil {
 96  1                  // cannot decode the response body, so it's not valid
 97  1                  violation := &errors.SchemaValidationFailure{
 98  1                          Reason:          err.Error(),
 99  1                          Location:        "unavailable",
100  1                          ReferenceSchema: string(renderedSchema),
101  1                          ReferenceObject: string(responseBody),
102  1                  }
103  1                  validationErrors = append(validationErrors, &errors.ValidationError{
104  1                          ValidationType:    helpers.ResponseBodyValidation,
105  1                          ValidationSubType: helpers.Schema,
106  1                          Message: fmt.Sprintf("%s response body for '%s' failed to validate schema",
107  1                                  request.Method, request.URL.Path),
108  1                          Reason:                 fmt.Sprintf("The response body cannot be decoded: %s", err.Error()),
109  1                          SpecLine:               1,
110  1                          SpecCol:                0,
111  1                          SchemaValidationErrors: []*errors.SchemaValidationFailure{violation},
112  1                          HowToFix:               errors.HowToFixInvalidSchema,
113  1                          Context:                string(renderedSchema), // attach the rendered schema to the error
114  1                  })
115  1                  return false, validationErrors
116  1          }
117     }
118
119     // no response body? failed to decode anything? nothing to do here.
120 13  if responseBody == nil || decodedObj == nil {
121  2          return true, nil
122  2  }
123
124     // create a new jsonschema compiler and add in the rendered JSON schema.
125  9  compiler := jsonschema.NewCompiler()
126  9  fName := fmt.Sprintf("%s.json", helpers.ResponseBodyValidation)
127  9  _ = compiler.AddResource(fName,
128  9          strings.NewReader(string(jsonSchema)))
129  9  jsch, _ := compiler.Compile(fName)
130  9
131  9  // validate the object against the schema
132  9  scErrs := jsch.Validate(decodedObj)
133 14  if scErrs != nil {
134  5          jk := scErrs.(*jsonschema.ValidationError)
135  5
136  5          // flatten the validationErrors
137  5          schFlatErrs := jk.BasicOutput().Errors
138  5          var schemaValidationErrors []*errors.SchemaValidationFailure
139 21          for q := range schFlatErrs {
140 16                  er := schFlatErrs[q]
141 21                  if er.KeywordLocation == "" || strings.HasPrefix(er.Error, "doesn't validate with") {
142  5                          continue // ignore this error, it's useless tbh, utter noise.
143                     }
144 22                  if er.Error != "" {
145 11
146 11                          // re-encode the schema.
147 11                          var renderedNode yaml.Node
148 11                          _ = yaml.Unmarshal(renderedSchema, &renderedNode)
149 11
150 11                          // locate the violated property in the schema
151 11                          located := schema_validation.LocateSchemaPropertyNodeByJSONPath(renderedNode.Content[0], er.KeywordLocation)
152 11
153 11                          // extract the element specified by the instance
154 11                          val := instanceLocationRegex.FindStringSubmatch(er.InstanceLocation)
155 11                          var referenceObject string
156 11
157 13                          if len(val) > 0 {
158  2                                  referenceIndex, _ := strconv.Atoi(val[1])
159  4                                  if reflect.ValueOf(decodedObj).Type().Kind() == reflect.Slice {
160  2                                          found := decodedObj.([]any)[referenceIndex]
161  2                                          recoded, _ := json.MarshalIndent(found, "", "  ")
162  2                                          referenceObject = string(recoded)
163  2                                  }
164                             }
165 20                          if referenceObject == "" {
166  9                                  referenceObject = string(responseBody)
167  9                          }
168
169 11                          violation := &errors.SchemaValidationFailure{
170 11                                  Reason:          er.Error,
171 11                                  Location:        er.KeywordLocation,
172 11                                  ReferenceSchema: string(renderedSchema),
173 11                                  ReferenceObject: referenceObject,
174 11                                  OriginalError:   jk,
175 11                          }
176 11                          // if we have a location within the schema, add it to the error
177 22                          if located != nil {
178 11
179 11                                  line := located.Line
180 11                                  // if the located node is a map or an array, then the actual human interpretable
181 11                                  // line on which the violation occurred is the line of the key, not the value.
182 14                                  if located.Kind == yaml.MappingNode || located.Kind == yaml.SequenceNode {
183  6                                          if line > 0 {
184  3                                                  line--
185  3                                          }
186                                     }
187
188                                     // location of the violation within the rendered schema.
189 11                                  violation.Line = line
190 11                                  violation.Column = located.Column
191                             }
192 11                          schemaValidationErrors = append(schemaValidationErrors, violation)
193                     }
194             }
195
196  5          line := 1
197  5          col := 0
198 10          if schema.GoLow().Type.KeyNode != nil {
199  5                  line = schema.GoLow().Type.KeyNode.Line
200  5                  col = schema.GoLow().Type.KeyNode.Column
201  5          }
202
203             // add the error to the list
204  5          validationErrors = append(validationErrors, &errors.ValidationError{
205  5                  ValidationType:    helpers.ResponseBodyValidation,
206  5                  ValidationSubType: helpers.Schema,
207  5                  Message: fmt.Sprintf("%d response body for '%s' failed to validate schema",
208  5                          response.StatusCode, request.URL.Path),
209  5                  Reason: fmt.Sprintf("The response body for status code '%d' is defined as an object. "+
210  5                          "However, it does not meet the schema requirements of the specification", response.StatusCode),
211  5                  SpecLine:               line,
212  5                  SpecCol:                col,
213  5                  SchemaValidationErrors: schemaValidationErrors,
214  5                  HowToFix:               errors.HowToFixInvalidSchema,
215  5                  Context:                string(renderedSchema), // attach the rendered schema to the error
216  5          })
217     }
218 14  if len(validationErrors) > 0 {
219  5          return false, validationErrors
220  5  }
221  4  return true, nil
222    }
# /src/github.com/pb33f/libopenapi-validator (fix-validate-security)>
```

</details>

However, I have confirmed that the coverage is indeed down, and I will try to determine the cause!

<details>

```console
# /src/github.com/pb33f/libopenapi-validator (main)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.636s  coverage: 51.9% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        0.944s  coverage: 91.0% of statements
ok      github.com/pb33f/libopenapi-validator/paths     1.142s  coverage: 36.8% of statements
ok      github.com/pb33f/libopenapi-validator/requests  0.681s  coverage: 36.0% of statements
ok      github.com/pb33f/libopenapi-validator/responses 1.307s  coverage: 40.7% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 1.588s  coverage: 33.9% of statements
# /src/github.com/pb33f/libopenapi-validator (main)> octocov

            main (b3f6e33)
----------------------------
  Coverage           96.0%

# /src/github.com/pb33f/libopenapi-validator (main)> git switch fix-validate-security
Switched to branch 'fix-validate-security'
# /src/github.com/pb33f/libopenapi-validator (fix-validate-security)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.430s  coverage: 50.0% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        1.199s  coverage: 90.7% of statements
ok      github.com/pb33f/libopenapi-validator/paths     0.968s  coverage: 35.5% of statements
ok      github.com/pb33f/libopenapi-validator/requests  1.395s  coverage: 33.7% of statements
ok      github.com/pb33f/libopenapi-validator/responses 0.485s  coverage: 39.6% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 0.726s  coverage: 34.3% of statements
# /src/github.com/pb33f/libopenapi-validator (fix-validate-security)> octocov

                fix-validate-security
                      (dd580b2)
--------------------------------------------
  Coverage                           95.5%
```

</details>

k1LoW commented:
It is already not covered two branches in the `main`.

ref: https://app.codecov.io/gh/pb33f/libopenapi-validator/blob/main/responses%2Fvalidate_response.go

<details>

```console
# /src/github.com/pb33f/libopenapi-validator (main)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.530s  coverage: 51.9% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        0.605s  coverage: 91.0% of statements
ok      github.com/pb33f/libopenapi-validator/paths     0.824s  coverage: 36.8% of statements
ok      github.com/pb33f/libopenapi-validator/requests  1.453s  coverage: 36.0% of statements
ok      github.com/pb33f/libopenapi-validator/responses 0.971s  coverage: 40.7% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 1.241s  coverage: 33.9% of statements
# /src/github.com/pb33f/libopenapi-validator (main)> octocov

            main (b3f6e33)
----------------------------
  Coverage           96.0%

# /src/github.com/pb33f/libopenapi-validator (main)> octocov ls-files
100.0% [    4/4] errors/error_utilities.go
100.0% [  50/50] errors/parameter_errors.go
100.0% [    6/6] errors/request_errors.go
100.0% [  18/18] errors/response_errors.go
 11.1% [    1/9] errors/validation_error.go
 78.3% [  18/23] helpers/operation_utilities.go
 92.4% [134/145] helpers/parameter_utilities.go
100.0% [  54/54] parameters/cookie_parameters.go
100.0% [  63/63] parameters/header_parameters.go
100.0% [    3/3] parameters/parameters.go
 99.3% [139/140] parameters/path_parameters.go
100.0% [  93/93] parameters/query_parameters.go
 98.6% [  69/70] parameters/validate_parameter.go
100.0% [  54/54] parameters/validate_security.go
100.0% [  93/93] parameters/validation_functions.go
 99.3% [143/144] paths/paths.go
100.0% [    3/3] requests/request_body.go
100.0% [  39/39] requests/validate_body.go
100.0% [  61/61] requests/validate_request.go
100.0% [    3/3] responses/response_body.go
100.0% [  52/52] responses/validate_body.go
 90.9% [  60/66] responses/validate_response.go
100.0% [  16/16] schema_validation/locate_schema_property.go
  0.0% [   0/24] schema_validation/openapi_schemas/load_schema.go
100.0% [  32/32] schema_validation/validate_document.go
100.0% [  83/83] schema_validation/validate_schema.go
 98.3% [119/121] validator.go
# /src/github.com/pb33f/libopenapi-validator (main)> git switch range-response-code
Switched to branch 'range-response-code'
# /src/github.com/pb33f/libopenapi-validator (range-response-code)> go test ./... -coverprofile=coverage.out -covermode=count
        github.com/pb33f/libopenapi-validator/helpers           coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/errors            coverage: 0.0% of statements
        github.com/pb33f/libopenapi-validator/schema_validation/openapi_schemas         coverage: 0.0% of statements
ok      github.com/pb33f/libopenapi-validator   0.432s  coverage: 51.9% of statements
ok      github.com/pb33f/libopenapi-validator/parameters        0.800s  coverage: 91.0% of statements
ok      github.com/pb33f/libopenapi-validator/paths     0.700s  coverage: 36.8% of statements
ok      github.com/pb33f/libopenapi-validator/requests  0.867s  coverage: 36.0% of statements
ok      github.com/pb33f/libopenapi-validator/responses 0.477s  coverage: 41.0% of statements
ok      github.com/pb33f/libopenapi-validator/schema_validation 0.595s  coverage: 33.9% of statements
# /src/github.com/pb33f/libopenapi-validator (range-response-code)> octocov

            range-response-code (a80bbb4)
-------------------------------------------
  Coverage                          96.0%

# /src/github.com/pb33f/libopenapi-validator (range-response-code)> octocov ls-files
100.0% [    4/4] errors/error_utilities.go
100.0% [  50/50] errors/parameter_errors.go
100.0% [    6/6] errors/request_errors.go
100.0% [  18/18] errors/response_errors.go
 11.1% [    1/9] errors/validation_error.go
 78.3% [  18/23] helpers/operation_utilities.go
 92.4% [134/145] helpers/parameter_utilities.go
100.0% [  54/54] parameters/cookie_parameters.go
100.0% [  63/63] parameters/header_parameters.go
100.0% [    3/3] parameters/parameters.go
 99.3% [139/140] parameters/path_parameters.go
100.0% [  93/93] parameters/query_parameters.go
 98.6% [  69/70] parameters/validate_parameter.go
100.0% [  54/54] parameters/validate_security.go
100.0% [  93/93] parameters/validation_functions.go
 99.3% [143/144] paths/paths.go
100.0% [    3/3] requests/request_body.go
100.0% [  39/39] requests/validate_body.go
100.0% [  61/61] requests/validate_request.go
100.0% [    3/3] responses/response_body.go
100.0% [  54/54] responses/validate_body.go
 90.9% [  60/66] responses/validate_response.go
100.0% [  16/16] schema_validation/locate_schema_property.go
  0.0% [   0/24] schema_validation/openapi_schemas/load_schema.go
100.0% [  32/32] schema_validation/validate_document.go
100.0% [  83/83] schema_validation/validate_schema.go
 98.3% [119/121] validator.go
# /src/github.com/pb33f/libopenapi-validator (range-response-code)>
```

</details>

k1LoW commented:
If you will allow me, I can contribute to the coverage by adding tests that are not related to this fix!

daveshanley commented:
> If you will allow me, I can contribute to the coverage by adding tests that are not related to this fix!

I would be most grateful! Thank would be very kind of you.

k1LoW commented:
Hi @daveshanley. I added tests for validate_response.go.

daveshanley commented:
@k1LoW You are a superstar! thank you so very much for all your contributions.

daveshanley reviewed (approved):
LGTM! Thank you!
